### PR TITLE
OAuth: Implement Azure AD provide

### LINKED ISSUE #21581: OAuth: Implement Azure AD provider
Currently you have to use Generic OAuth configuration to use Azure AD, but it's a bit limited. Suggest to Implement specific Azure AD OAuth provider.

Related issues: 
- OpenID unique user identifier #17635
- Add support for teams or orgs in Azure AD OAuth #15670
- id_token signature validation #15312

### PR DESCRIPTION
**Pull request by twendt**
**What this PR does / why we need it**:
Currently Azure AD can be used with the generic oauth. This does not support roles and besides that I realized that it does not seem to validate the JWT token. Therefore I decided to implement a separate oauth for Azure AD and included the token validation by using the openid well known configuration to find the signing key.

Closes #21581

**Special notes for your reviewer**:
This is my first PR, so I might need some help to get in shaped.

### CODE REVIEW

**Comment by marefr** [comment]
@twendt thanks. Looks interesting. We're planning to review this soon. Until then, could you please fix the following:

**Linting issues:**
You can run the linting locally using `make golangci-lint`
```
lint via golangci-lint
pkg/login/social/azuread_oauth.go:16:2: field `hostedDomain` is unused (unused)
	hostedDomain   string
	^
pkg/login/social/openid_config.go:65:17: func `(*Token).getJwksUri` is unused (unused)
func (t *Token) getJwksUri() (string, error) {
                ^
pkg/login/social/token.go:13:2: field `jwksCert` is unused (unused)
	jwksCert      string
	^
pkg/login/social/openid_config.go:55:12: SA4006: this value of `err` is never used (staticcheck)
	response, err := HttpGet(client, wellKnownUrl)
	          ^
pkg/login/social/openid_config.go:70:12: SA4006: this value of `err` is never used (staticcheck)
	response, err := HttpGet(http.DefaultClient, t.wellKnownUrl)
```

**Dependency/build problem:**
Could you please use the existing jwt library in grafana, i.e. https://gopkg.in/square/go-jose.v2 / https://github.com/square/go-jose/tree/v2
```
Executing 'time go install ./pkg/cmd/grafana-server'
build github.com/grafana/grafana/pkg/cmd/grafana-server: cannot load github.com/dgrijalva/jwt-go: open /go/src/github.com/grafana/grafana/vendor/github.com/dgrijalva/jwt-go: no such file or directory
```

**Comment by twendt** [comment]
Thanks for looking into this.

Before I start the refactoring I have an important question. Can I trust the id_token that I get passed in? If that is the case, then there is no point for me to do the validation steps. I think I made a wrong assumption here and that it does not have to be validated.

The main point of this PR is to enable roles. They don't work with the generic oauth provider as the roles are available from the id_token. The roles are however evaluated only from the response body of the apiURL request. This is not queried as there is an id_token available. Even if it would be queried then it would not include the roles.

**Comment by marefr** [comment]
@twendt currently there are no validation of tokens happening. There are an open feature request for it #15312.

**Comment by twendt** [comment]
How should we proceed with this PR then? I would be fine with #21074. For future changes however it might also be a good idea to have a separate implementation for Azure AD instead of implementing too many special cases into the generic oauth.

**Comment by marefr** [comment]
Probably a good idea to add azure ad as a specific provider if it provides something more than just role mapping (which #20300 will do). For example if it makes it easier for end users to configure azure ad having it as a separate provider it could make sense. Regarding token validation that should definitely be handled in a separate PR.

**Comment by marefr** [comment]
@twendt sorry for keeping you waiting. We received this issue #21366 a while ago which could only be resolved properly if azure ad had its own provider implemented. 

Please let me know if you plan to continue working on this PR and/or add support for azure ad oauth, otherwise I suggest we close this. Thanks

**Comment by twendt** [comment]
I will try to find the time next week.

**Comment by marefr** [comment]
@twendt just opened a specific issue for supporting Azure AD as a specific OAuth provider #21581 

Regarding id_token signature validation, there's an existing feature request opened #15312

**Comment by dekimsey** [comment]
This is using the v1 oauth2 implementation is this intentional? v2.0 is actually OIDC compliant (supposedly) and may be a better target. IIRC there were some scope implementation bits that differ here.

**Comment by marefr** [comment]
@dekimsey thanks. Think you can add your thoughts/suggestions to #21581?

**Comment by twendt** [comment]
@dekimsey I have changed it to v2.

File: docs/sources/auth/azuread.md
Comment (oddlittlebird) [inline docs/sources/auth/azuread.md]:
```suggestion
4. Click **Endpoints**.
```

File: docs/sources/auth/azuread.md
Comment (oddlittlebird) [inline docs/sources/auth/azuread.md]: Please add more of an intro here to give the reader context. Tell the reader what to expect. Imagine you are looking at search results and ask yourself if this paragraph gives you enough information to know if the page contains the information you need. Right now, I do not think it does.

File: docs/sources/auth/azuread.md
Comment (oddlittlebird) [inline docs/sources/auth/azuread.md]:
```suggestion
> Note: Ensure that the [root_url]({{< relref "../installation/configuration/#root-url" >}}) in Grafana is set in your Azure Application Reply URLs (App -> Settings -> Reply URLs)
```

File: docs/sources/auth/azuread.md
Comment (oddlittlebird) [inline docs/sources/auth/azuread.md]: Per the Documentation Style Guide, avoid having only one child heading. There should be at least two h2 headings under an h1, or none.

File: docs/sources/auth/azuread.md
Comment (oddlittlebird) [inline docs/sources/auth/azuread.md]: Per the Documentation Style Guide, all headings should be in sentence case. 

Also, this topic looks like a task, so I suggest starting it with a verb. Something like Set up Azure AD OAuth2 authentication.

File: docs/sources/auth/azuread.md
Comment (oddlittlebird) [inline docs/sources/auth/azuread.md]: Where do I set up this configuration? It looks like you want the reader to copy the content below and paste it somewhere, but I do not know where. Please clarify.

**Review by oddlittlebird — DISMISSED** [review]
azuread.md has some of the clearest instructions I have seen submitted. Well done! I have some suggestions to improve them and make them more consistent with Grafana style guidelines.

Please review the [Documentation Style Guide](https://github.com/grafana/grafana/blob/master/contribute/style-guides/documentation-style-guide.md).

Per the Microsoft Style Guide, clickable elements should be bold, not surrounded in quotation marks.

Also, please be sure to put periods at the end of sentences, including ordered and unordered list items.

URL should be capitalized.

File: docs/sources/auth/azuread.md
Comment (twendt) [reply docs/sources/auth/azuread.md]: I changed all the clickable items to be bold.

File: docs/sources/auth/azuread.md
Comment (twendt) [reply docs/sources/auth/azuread.md]: I added some more context.

File: docs/sources/auth/azuread.md
Comment (twendt) [reply docs/sources/auth/azuread.md]: Changed

File: docs/sources/auth/azuread.md
Comment (twendt) [reply docs/sources/auth/azuread.md]: I changed it to a h1, but I am not sure if that was correct.

File: docs/sources/auth/azuread.md
Comment (twendt) [reply docs/sources/auth/azuread.md]: I added the link to the Grafana configuration file.

**Comment by twendt** [comment]
Can someone help me out on how to get an Azure icon into the login button?

**Comment by marefr** [comment]
@twendt you'll need to populate it here https://github.com/grafana/grafana/blob/249ae05281190ffcf8079a61cc7d791225f1673a/public/app/core/components/Login/LoginServiceButtons.tsx#L7-L38

Most of the providers use font awesome for the icon, but seems there's no one for azure. You can see here how it's possible to overide using a custom icon:
https://github.com/grafana/grafana/blob/249ae05281190ffcf8079a61cc7d791225f1673a/public/sass/components/_buttons.scss#L210-L217

File: pkg/login/social/azuread_oauth.go
Comment (alexanderzobnin) [inline pkg/login/social/azuread_oauth.go]: `unique_name` as mentioned in [Azure AD docs](https://docs.microsoft.com/en-us/azure/active-directory/develop/id-tokens#payload-claims) is only issued in v1.0 ID tokens. I think it makes sense to use `preferred_username` intead as a backup field for getting email.

File: pkg/login/social/azuread_oauth.go
Comment (alexanderzobnin) [inline pkg/login/social/azuread_oauth.go]: As mentioned above, it's better to replace `unique_name` with `preferred_username` since it's issued in v2.0 tokens (or add `preferred_username` as additional field to check).

**Review by alexanderzobnin — CHANGES_REQUESTED** [review]
Good job, @twendt ! Only some notes about claims.
I configured Azure AD OAuth application with docs you provided and couldn't log in because of error:
```
Error getting user info: No email found in access token
```
This happened because none of `email`, `unique_name` or `upn` claims presented in ID token returned by application. I parsed ID token and found email returned in `preferred_username` claim. Also, adding optional `upn` claim at the _Token configuration_ page solves a problem (since you use `upn` as a backup claim for searching email).
My suggestion is use `preferred_username` claim for getting email if `email` claim is empty, because it's a claim [returned by default](https://docs.microsoft.com/en-us/azure/active-directory/develop/id-tokens#payload-claims) in v2.0 tokens. Also, it makes sense I think to add info how to add `upn` claim to the ID token since it's optional claim.
Otherwise, it works fine and authentication passed without any other issues.

File: pkg/login/social/azuread_oauth.go
Comment (alexanderzobnin) [inline pkg/login/social/azuread_oauth.go]: I there a reason to check string length, not just check `claims.Upn != ""`?
```suggestion
		if claims.Upn != "" {
```

File: pkg/login/social/azuread_oauth.go
Comment (alexanderzobnin) [inline pkg/login/social/azuread_oauth.go]: It seems, this function doesn't use `s *SocialAzureAD`, so maybe it makes sense to convert it to regular func, not a method on `SocialAzureAD` type, just to be consistent.

File: pkg/login/social/azuread_oauth.go
Comment (twendt) [reply pkg/login/social/azuread_oauth.go]: I will change it to `preferred_username`.

File: pkg/login/social/azuread_oauth.go
Comment (twendt) [reply pkg/login/social/azuread_oauth.go]: You are right. I tested with v1 and v2. I will change this to `preferred_username`. Are you not getting the email even when `email` is added to the scopes? I thought that is standard for openid connect.

File: pkg/login/social/azuread_oauth.go
Comment (twendt) [reply pkg/login/social/azuread_oauth.go]: I will change this, there was no special reason for it.

File: pkg/login/social/azuread_oauth.go
Comment (twendt) [reply pkg/login/social/azuread_oauth.go]: Good point, I will change that as well.

File: pkg/login/social/azuread_oauth.go
Comment (alexanderzobnin) [reply pkg/login/social/azuread_oauth.go]: I didn't get email even if it's added in scopes. But I think it's related more to my account, not to grafana. I use google account for signing in, so maybe email is not set properly. I'll try to add email and check it again.

File: conf/defaults.ini
Comment (alexanderzobnin) [inline conf/defaults.ini]: This should be in `sample.ini` as well I guess.

File: public/app/core/components/Login/LoginServiceButtons.tsx
Comment (alexanderzobnin) [inline public/app/core/components/Login/LoginServiceButtons.tsx]: I'm not 100% sure about button label. @torkelo @daniellee @xlson what do you think - which name should we use - Microsoft, Azure, Azure AD, Microsoft Azure, etc?

**Review by alexanderzobnin — COMMENTED** [review]
Looks good to me! Just few minor notes. And the main question left is how name a login button :)

File: public/app/core/components/Login/LoginServiceButtons.tsx
Comment (twendt) [reply public/app/core/components/Login/LoginServiceButtons.tsx]: I name it Microsoft due to this guide https://docs.microsoft.com/en-us/azure/active-directory/develop/howto-add-branding-in-azure-ad-apps I also got the icon from there.

File: pkg/login/social/azuread_oauth.go
Comment (twendt) [reply pkg/login/social/azuread_oauth.go]: Changed

File: pkg/login/social/azuread_oauth.go
Comment (twendt) [reply pkg/login/social/azuread_oauth.go]: Changed

File: pkg/login/social/azuread_oauth.go
Comment (twendt) [reply pkg/login/social/azuread_oauth.go]: Changed

File: conf/defaults.ini
Comment (twendt) [reply conf/defaults.ini]: Sorry, I haven't seen that file before. I added the section there as well.

File: public/app/core/components/Login/LoginServiceButtons.tsx
Comment (alexanderzobnin) [reply public/app/core/components/Login/LoginServiceButtons.tsx]: Ok, `Sign in with Microsoft` looks good thus.
>DON’T expose end users to the Azure or Active Directory brands. It’s ok however to use these terms with developers, IT pros, and admins.

The only question are all Grafana users _developers, IT pros, and admins_ so we can use Azure name :)
But anyway, from the Microsoft branding guide `Sign in with Microsoft` looks better and quite unambiguous.

**Review by alexanderzobnin — APPROVED** [review]
LGTM! Let's wait for comments regarding sign in button name, otherwise looks good.

File: public/app/core/components/Login/LoginServiceButtons.tsx
Comment (xlson) [reply public/app/core/components/Login/LoginServiceButtons.tsx]: `Sign in with Microsoft` is in line with the rest of the ui. We already have `Sign in with Google` and `Sign in with Grafana.com`

File: docs/sources/auth/azuread.md
Comment (xlson) [reply docs/sources/auth/azuread.md]: I think the best way go forward here is to use H1 for the header and then split the article into two parts. One for the guide of setting up the application on the azure side and one for enabling oauth in grafana. We should use H2 for those headers. If you want you can change that, or we can fix it after merging.

Comment (alexanderzobnin) [reply docs/sources/auth/azuread.md]: Ok, it looks like a minor thing, so we can merge it now and then update docs styles.

**Pull request merged**
